Init checkin of client changes

hongdai opened the pull request:
* Allow Test property ServiceUri.
* Note that we in the self hosted case, we only allow localhost, not multiple
service instances(localhost/wcfservice1).
* Change all endpoints point to the new services.

zhenlan commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+    //Install Root certificate, this is required for all https test
+    private static void EnsureRootCertificateInstalled()
+    {
+        lock (rootCertLock)
```
NIT: null check of `rootCert` before taking a lock.

iamjasonp replied:
Do we really need a lock here? 

1. The call to `serviceProxy.GetRootCert(false)` should be idempotent during one test run. 
2. BridgeClientCertificateManager will take the appropriate locks and do the appropriate checks to ensure the cert is only installed once per system

zhenlan commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+    {
+        EnsureRootCertificateInstalled();
+        lock (clientCertLock)
+        {
```
NIT: null check of `clientCert` before taking a lock.

iamjasonp replied:
ditto on whether or not we need to take a lock for obtaining the clientCertificate.

zhenlan commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+
+    private static string GetServiceHostName()
+    {
+        return TestProperties.GetProperty(TestProperties.BridgeHost_PropertyName);
```
Is this supposed to be changed to `ServiceUri_PropertyName`?

hongdai replied:
Nope. Bridge host name is just the service full qualified host name. It's used by some of the tests.

zhenlan replied:
So do you still need to pass BridgeHost in addition to ServiceUri in new infrastructure? If this is the case, I think we should parse host name (if that's what you need) from ServiceUri instead (trim everything after first '/' or '\').

---
In reply to: [59492053](https://github.com/dotnet/wcf/pull/1002#discussion_r59492053) [](ancestors = 59492053)

roncain replied:
+1 about parsing if possible for the tests.  If we need 2 properties, it should no longer be called "BridgeHost"

---
In reply to: [59492624](https://github.com/dotnet/wcf/pull/1002#discussion_r59492624) [](ancestors = 59492624,59492053)

hongdai replied:
Some tests need fully qualified server host name even it's localhost. So parsing is difficult. Another way is to get from the server, I have the method in the IUtil interface.

zhenlan commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+    {
+        get
+        {
+            if (TestProperties.GetProperty(TestProperties.ServiceUri_PropertyName).ToLower()==("localhost"))
```
Just to point out technically, ServiceUri==localhost does not necessarily mean the service is not IIS hosted. This will work in our current plan of lab/CI setup. Hopefully we can find a better way in the future, or even better if we don't have to differentiate IIS or self hosted at all.

hongdai replied:
I originally had a test property called IISHosted. If there is a requirement for "localhost" with IIS hosted, we could add this test property back. The drawback is one more test property to set.

zhenlan replied:
Agree. Let's not invent something new yet. If needed, we can do smarter detection.

iamjasonp replied:
NIT: use `string.Compare(TestProperties.GetProperty(TestProperties.ServiceUri_PropertyName, "localhost", StringComparison.OrdinalIgnoreCase)`

zhenlan commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+            switch (protocol)
+            {
+                case "http":
+                    builder.Port = int.Parse(TestProperties.GetProperty(TestProperties.BridgeHttpPort_PropertyName));
```
Can we just use the same ports (and the standard ports) for both IIS hosted and self hosted services? It will make machine setup much easier. We can do this as a follow up item.

hongdai replied:
I don't see a blocking issue to use the same ports. I'll open a follow up issue.

iamjasonp replied:
If someone's dev machine has IIS running and we try to selfhost on port 80 and 443, isn't that an issue?

roncain replied:
Not blocking: All test properties with Bridge need to be renamed at some point.

zhenlan commented:
Other than some NITs for your consideration this LGTM.

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/src/testproperties.props:
```diff
@@ -105,7 +105,11 @@
-  
+
+    <PropertyGroup Condition="'$(ServiceUri)' == ''">
+      <ServiceUri >localhost</ServiceUri >
```
1. Is this necessarily correct to have the ServiceUri be this by default? 
2. Should we split out the hostname and basepath like we talked about earlier?

roncain replied:
Micro nit: remove extra spaces in "<ServiceUri >"

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/ServiceInterfaces.cs:
```diff
@@ -502,3 +502,13 @@ public interface IWcfAspNetCompatibleService
+    byte[] GetClientCert(bool exportAsPem);
+
+    [OperationContract]
+    byte[] GetRootCert(bool exportAsPem);
```
NIT: spell out the name, e.g., `GetServiceRootCertificate`

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/ServiceInterfaces.cs:
```diff
@@ -502,3 +502,13 @@ public interface IWcfAspNetCompatibleService
+public interface IUtil
+{
+    [OperationContract]
+    byte[] GetClientCert(bool exportAsPem);
```
NIT: spell out name, e.g. `GetMachineClientCertificate` or `GetUserClientCertificate` (there are two types of client certs we may need to use)

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+
+                    if (rootCert == null)
+                    {
+                        //throw
```
Delete. It's super obvious that you're throwing an exception here.

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+        lock (rootCertLock)
+        {
+            if (rootCert == null)
+            {
```
We call `EnsureRootCertificateInstalled()` in multiple places. 

If the root cert fails the first time, by https://github.com/dotnet/wcf/pull/1002/files#r59512010 bullet 1, it should fail every time we call this. Hence, we should use some kind of fail fast mechanism to fail all further calls to EnsureRootCertificateInstalled() in a particular run. 

We can't detect this condition currently because we are only keying off if `rootCert == null`

roncain replied:
Not blocking:  For future consideration we might let the first test hit and report the failure and then set a value so the remaining [ConditionalFact] tests depending on the root cert show as Skipped due to no root cert. This will satisfy @iamjasonp 's suggestion not to attempt doing it again if it fails once.

roncain replied:
Not blocking: note to capture for future thinking -- perhaps we should make the [ConditionalFact] for "Root_Certificate_Installed" be the one to call EnsureRootCertificateInstalled.  This would eliminate the need to try to keep both Endpoint.cs and the tests in sync regarding what they depend on.

---
In reply to: [59565735](https://github.com/dotnet/wcf/pull/1002#discussion_r59565735) [](ancestors = 59565735)

roncain replied:
FYI: I opened #1007 to consider moving this logic to [ConditionalFact].

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+                {
+                    factory = new ChannelFactory<IUtil>(basicHttpBinding, new EndpointAddress(ServiceUtil_Address));
+                    serviceProxy = factory.CreateChannel();
+                    rootCert = new X509Certificate2(serviceProxy.GetRootCert(false));
```
NIT: consider using named params here to spell out what's false here - `GetRootCert(exportAsPem: false)`

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+                }
+                finally
+                {
+                    ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)serviceProxy, factory);
```
I'm a _little_ weirded out that we're using a "ScenarioTestHelper" in a file where we're not doing any scenario tests. 
Not blocking.

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+    }
+
+    //Install client certificate
+    private static void EnsureLocalClientCertifciateInstalled()
```
Typo: EnsureLocalClientCertif**ciate**Installed

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+    private static X509Certificate2 clientCert;
+    private static object rootCertLock = new object();
+    private static object clientCertLock = new object();
+    //Install Root certificate, this is required for all https test
```
NIT: not only all https tests, but all transport security tests

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+                    builder.Scheme = "https";
+                    break;
+                case "net.tcp":
+                    builder.Port = int.Parse(TestProperties.GetProperty(TestProperties.BridgeTcpPort_PropertyName));
```
NIT: could use ushort.Parse instead to detect errors in passed-in port numbers.

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+
+        return builder.Uri;
+    }
+    public static string GetEndpointAdddress(string endpoint, string protocol = "http")
```
My opinion on this: given the nature of the parameters being passed, it would read much more clearly in subsequent usages if you explicitly called out the protocol ("http") rather than made it an optional argument. 

For example 
`GetEndpointAdddress("WebSocketHttpDuplexTextBuffered.svc//WebSocketHttpDuplexTextBufferedResource", protocol: "ws");`

reads better to me as 

`GetEndpointAdddress("ws", "WebSocketHttpDuplexTextBuffered.svc//WebSocketHttpDuplexTextBufferedResource");`

Similarly, 
`GetEndpointAdddress("NetHttp.svc//NetHttp")`

doesn't call out the protocol unless i look up the call to GetEndpointAddress, but 

`GetEndpointAdddress("http", "NetHttp.svc//NetHttp")` 

does

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+    }
+
+    private static Uri BuildBaseUri(string protocol)
+    {
```
I would Contract.Assert(protocol != null) here

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+    {
+        var builder = new UriBuilder();
+        builder.Host = TestProperties.GetProperty(TestProperties.ServiceUri_PropertyName);
+        builder.Scheme = protocol;
```
This should be moved into the switch statement. The `default:` case should throw.

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+                    builder.Port = int.Parse(TestProperties.GetProperty(TestProperties.BridgeTcpPort_PropertyName));
+                    break;
+                default:
+                    break;
```
The `default:` case should throw, since you'd be trying to create a Uri without a protocol.

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+                    clientCert = new X509Certificate2(serviceProxy.GetClientCert(false), "test", X509KeyStorageFlags.PersistKeySet);
+                    if (clientCert == null)
+                    {
+                        //throw
```
elide. It's obvious you're throwing

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -53,390 +187,394 @@ public static string HttpProtocolError_Address
         {
-            BridgeClientCertificateManager.InstallRootCertificateFromBridge();
-            return BridgeClient.GetResourceAddress("WcfService.TestResources.BasicAuthResource");
+            EnsureLocalClientCertifciateInstalled();
```
I think you need the root cert here not client cert.

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -53,390 +187,394 @@ public static string HttpProtocolError_Address
-            BridgeClientCertificateManager.InstallRootCertificateFromBridge();
-            return BridgeClient.GetResourceAddress("WcfService.TestResources.HttpsNtlmResource");
+            EnsureLocalClientCertifciateInstalled();
+            return GetEndpointAdddress("HttpsNtlm.svc//https-ntlm.svc//");
```
Why the extra slashes at end?

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -53,390 +187,394 @@ public static string HttpProtocolError_Address
         {
-            BridgeClientCertificateManager.InstallRootCertificateFromBridge();
-            return BridgeClient.GetResourceAddress("WcfService.TestResources.HttpsNtlmResource");
+            EnsureLocalClientCertifciateInstalled();
```
Root vs. client cert

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -53,390 +187,394 @@ public static string HttpProtocolError_Address
         {
-            return BridgeClient.GetResourceHostName("WcfService.TestResources.TcpRevokedServerCertResource");
+            return GetServiceHostName();
         }
```
Question only: how are we going to be testing this scenario?

hongdai replied:
IIS only has restriction on https port. For TCP port, we can have as many certs as we want.

roncain commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs:
```diff
@@ -29,7 +29,7 @@ public static class BridgeClientCertificateManager
         private const string ClientCertificateSubject = "WCF Client Certificate";
```
Not blocking -- remove "Bridge" from file and class names at some point.

roncain commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+    }
+
+    public static string ServiceUtil_Address
+    {
```
Not blocking: we need a better name for the service interface, class and endpoint name.  Maybe CertificateService_Address.

roncain commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+    public static string ServiceUtil_Address
+    {
+        get { return GetEndpointAdddress("Util.svc//Util"); }
+    }
```
Not blocking: (but the debt is increasing if we don't fix these soon) -- need a better name, like CertificateService.

roncain commented on src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs:
```diff
@@ -1,49 +1,183 @@
+            {
+                BasicHttpBinding basicHttpBinding = new BasicHttpBinding();
+                ChannelFactory<IUtil> factory = null;
+                IUtil serviceProxy = null;
```
Not blocking: I'll stop pointing out this name needs to be fixed, preferrably before its use spreads too far.

hongdai commented:
Since I have one LGTM, I will go ahead merge to unblock the progress.

I have opened a issue https://github.com/dotnet/wcf/issues/1006 to track addressing the comments in the future iterations.

iamjasonp commented:
No objection to merge now and fix later... I think one of my messages didn't get through. Sorry if that wasn't clear :smile:
